### PR TITLE
Fix issue 1485

### PR DESCRIPTION
**Pull request by LeoHsiao1**

### CODE REVIEW

**Comment by LeoHsiao1** [comment]
In this PR, I use the `Executer` class to execute the test commands, making the environment variables of `tests\bugfixes\github\test_pr1475_*.py` consistent with the other bash test cases.

**Comment by LeoHsiao1** [comment]
For the `runShell()` function you imagined last night, My thoughts are as follows:
If it were before we rewrote the bash test, `runShell()` would be attractive and easy to port shell commands to Python.
But now, new test cases can be implemented using only Python code. 
We can write the test cases as Dan does, or we can write the test cases by calling tests/bash_tests/utils.py .
For example, `exiv2 | grep MIME | cut -d: -f 2` could be written as follows:
```py
for line in Executer('exiv2').stdout:
    if 'MIME' in line:
        if line.split(':')[1] != 'image/heif':
            print('FAILED!')
            return 1
```
If it's too wordy, I can optimize it to:
```py
if Executer('exiv2).grep('MIME').cut(d=':', f=2) != 'image/heif':
    print('FAILED!')
    return 1
```
All in all, these test cases can be implemented using only Python code, and are probably more convenient and understandable than shell commands.

**Comment by clanmills** [comment]
This is an excellent suggestion.  I enthusiastically buy what you've said.  We want to bury bash and platform utilities like cut and grep.

The Exiv2 tools are: cmake+compiler+python.  The user can install dependant libaries (zlib and expat) with conan or platform tools.  This is a very nice way to proceed and takes advantage of your "pythonic tools" such as cut/diff/grep.

I'll pull down and review your code.

**Review by clanmills — APPROVED** [review]
This looks GREAT.  Very good work.  You've connected everything into a package.  Very good work.

**Comment 1**
In tests such as test_pr1475_HIV.py that require enable_bmff=1, can we simply the code to a one liner:

```python
system_tests.skip('requires enable_bmff=1') if system_tests.BT.verbose_version('enable_bmff') != '1'
```
And it would be really nice, if it skip() emitted 's' on the ribbon and only throw up messages in the terminal when --verbose is set.  I've totally invented system_tests.skip().  I don't know how Dan's code works.

**Comment 2**
I was really impressed by what you said about emulating bash in python.  Can you think about writing something about this in tests/README-TESTS.md  If you write something, I'll revise your English.

**Comment 3**
It would be good to merge the directories test/ and /tests/.  Let's do that in Exiv2 v0.28 which will be on branch 'main'.  We'll create 'main' after v0.27.4 ships.  So, although the test suite no longer executes bash scripts, let's leave the scripts in v0.27.4.  We will remove them in v0.28.

**Comment by clanmills** [comment]
**Comment 4**
In suite.conf, the default for HTTP_PORT is 
```
exiv2_port: 12760
```

It should be 12761 and 12762 for MinGW and Cygwin.  See test/functions.source.  On a Windows builder, it's possible for  msvc, cygwin and mingw to be running the test suite at the same time.  I want them to use different ports to avoid port contention.  The platform is known in verbose_version().get('platform').  Maybe the simple fix is to have code where you run the HTTP server to say:

port += 1 if verbose_version().get('platform') == 'mingw'
port += 1 if verbose_version().get('platform') == 'cygwin'

**Comment by LeoHsiao1** [comment]
Your Comment 4 has now been resolved.

**Comment by LeoHsiao1** [comment]
Your Comment 2 has now been resolved.

**Review by clanmills — COMMENTED** [review]
This looks good, young Mr Lion.  We're going to visit family this morning and I'll look closely at this after lunch.  Several points immediately hit my eye.

1.  Is there a need for the utils directory.  Can common_code.py not go into directory bash_tests/?
2. Is common_code.py a good name.  What about emulate_utils.py?
3. Be more assertive when you state something.  For example: `Here are some common code that are poorly coupled with test cases.` is very negative and provides no useful information.  Say what it does.
4. `thus all test files are Python files` should be `thus all test scripts are written in python`
5. `The test suite’s configuration file should have the following form:` should be `The suite's configuration file has the following form:`
6. I see you haven't updated bugfixes/github/test_pr1475_HIF.py.  What are your thoughts about that?
7. Can you provide on-line help?  For example:
```bash
$ python3
Python 3.7.2 (v3.7.2:9a3ffc0492, Dec 24 2018, 02:44:43) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import system_tests
>>> help(system_tests)
....
>>> help(system_tests.BT)
Help on package utils:

NAME
    utils

PACKAGE CONTENTS
    common_code
    configs
    coupled_code

DATA
    log = <utils.common_code.Log object>

FILE
    /Users/rmills/gnu/github/exiv2/0.27-maintenance/tests/utils/__init__.py

(END)
```

I think you have something very useful here that could be adopted by other projects.  The nice thing about bash is that it's quick and easy to use.  The bad thing is that Windows users mostly dislike stuff that smells of UNIX.  You and Dan have constructed something here that provides the convenience of bash from python.

**Comment by LeoHsiao1** [comment]
Your Comment 1 has now been resolved.
I found that in unittest, throwing this exception skips the test case:
```py
raise unittest.SkipTest('Why skip it')
```

**Comment by LeoHsiao1** [comment]
- I plan to put all the common functions and classes in the `utils` directory. It may be called by all test cases and is therefore not a subdirectory of `bash_tests`.
- The previous utils.py file was too large, so I split it into several files. Please draw on your extensive experience to come up with some suitable file name. I'm not sensitive to the word `common`.
- I've just added some simple instructions to utils.
- As the weekend draws to a close, I'm going to get a hamburger.

**Comment by clanmills** [comment]
Enjoy your hamburger.  Today is Mother's Day in England.  We went for a walk with our son and his wife.  I'll look at your code after lunch.

**Comment by clanmills** [comment]
@LeoHsiao1 I think there is still quite a lot of work required here to get this to the standard that I have in mind.

RC1 is well ahead of schedule.  Lets work together this week to get everything solid for RC1.  How much time do you have available and can you work on this next weekend?

I'll take responsibility for README-TESTS.md.  I updated about a quarter in about one hour.  By the end of the week, this document will be really solid.

As always I have a number of small issues.

1.  using unittest Skiptest.

Please remove the vv variable in bugfixes/github/test_pr1475*.py scripts:

```python
if system_tests.BT.verbose_version().get('enable_bmff') != '1':
    raise unittest.SkipTest('*** requires enable_bmff=1 ***')
```

I feel there should be a beautiful one liner, but I can't get the following to compile.  Is 'raise' a reserved word in python?

```python
raise unittest.SkipTest('*** requires enable_bmff=1 ***') if system_tests.BT.verbose_version().get('enable_bmff') != '1'
```
Ugly one liner:
```python
if system_tests.BT.verbose_version().get('enable_bmff') != '1':; raise unittest.SkipTest('*** requires enable_bmff=1 ***') 
```

2. Filenames:

configs.py => config.py
common_code.py   **Good Name**
coupled_code.py  => executer.py _**and**_ output.py _**and**_ tests.py

3. Other stuff

There will be lots of little things to fix.  It's very important that you deal with the open issue concerning `$ make bash_tests` etc.  #1274

So before I list every detail, please deal with #1274 and the matters I've listed above.

------------------

_**Something totally different.**_

I'm very pleased with the guys in Europe who have worked on Exiv2 v0.27.4.  Milos, Peter, Christoph, Luis, Dan and Alex.    We're thinking about a mountain break weekend in Bavaria in September.   There are several others who have contributed in the past and might join us.

Alex and I are discussing 3 days in a cabin in the mountains to hike, talk, eat, booze and simply "hang out".  Is it possible for you to join us?  Possible or impossible?

**Comment by clanmills** [comment]
I've spent about 3 hours on README-TESTS.md this morning. It's going to take 2-3 hours every day this week to bring this document to the English standard of our other documents.

Please have a look at the bottom of the document.  I haven't understood the architecture of system\_tests.BT.  Please give me an outline of what should be written and I will take care of the English.

I'm going start to work on Exiv2 v0.27.4 RC1 Preview and will probably publish that tomorrow.

Let's aim to publish Exiv2 v0.27.4 RC1 next week with everything in the test suite working and documented correctly.  In particular, I'd like to see #1274 fixed and closed for RC1.

**Comment by clanmills** [comment]
@LeoHsiao1 This is impossibly complicated.  How does it work?  Let's take something simple such as running the command `$ exiv2 --verbose --version` using BT.

```bash
1238 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance/tests $ python3
Python 3.7.2 (v3.7.2:9a3ffc0492, Dec 24 2018, 02:44:43) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> import unittest
>>> import re
>>> import system_tests
>>> from system_tests import BT
>>> os.makedirs(BT.Config.tmp_dir, exist_ok=True)
>>> os.chdir(BT.Config.tmp_dir)
>>> BT.Config.init()
>>> BT.Executer('exiv2 -vV')
<utils.coupled_code.Executer object at 0x7fe23046fda0>
>>> 
```

I expect it to work effortlessly like this:

```bash
$ cd tests
$ python
>>> import system_tests
>>> system_tests.BT.Executer('exiv2 -vV')
exiv2 0.27.4.9
exiv2=0.27.4
platform=apple
compiler=Clang
...
```

You have all the code.  However the documentation and packaging are both poor.  I'm working on the documentation.  Please work with me to fix the packaging.  Together we can make a success of your work.

We are now 2 weeks past "Code Complete" which was 2021-02-29.  If your code cannot be included in v0.27.4 RC1 (next week), it's unlikely to ship in v0.27.4.

**Comment by LeoHsiao1** [comment]
Enjoy your hike in the woods!
I thought about your suggestion, and after I ate dinner, I put in these commit.

- `raise` is indeed a keyword in Python, similar to `throw` in C++.
    This line is too long:
    ```py
    if system_tests.BT.verbose_version().get('enable_bmff') != '1':; raise unittest.SkipTest('*** requires enable_bmff=1 ***') 
    ```
    I prefer to split it into two lines:
    ```py
    if system_tests.BT.verbose_version().get('enable_bmff') != '1':
        raise unittest.SkipTest('*** requires enable_bmff=1 ***')
    ```

- I updated the directory structure and documentation for `tests/utils`.
  When I was learning English, I often had trouble remembering the plural forms of nouns.
  Sometimes I write `test`, `config`. And sometimes I write `tests`, `configs`.

- Another tip: Python3 reads script files in UTF-8 by default, so you don't need to add `# -*- coding: utf-8 -*-` as the first line of the file.

**Comment by LeoHsiao1** [comment]
The original `Executer` was used this way:
```py
>>> from system_tests import BT
>>> BT.Executer('echo Hello').stdout
Hello
```
I just improved it and now you can use it like this:
```py
>>> from system_tests import BT
>>> BT.Executer('echo Hello')
Hello
>>> BT.Executer('exiv2 -vV')
exiv2 0.27.4.9
exiv2=0.27.4
platform=linux
..
```

**Comment by clanmills** [comment]
Please change 'Executer' to 'Run' _or Exec_.  It's easier to type and no confusion with the English word Executioner: https://en.wikipedia.org/wiki/Executioner.

**Comment by LeoHsiao1** [comment]
- If the command you executed exists in the `Config.bin_dir` directory, it will be used and converted to an absolute path.
  Otherwise, it is treated as a normal command, subject to factors such as PATH.
- The environment variables are set as follows in `BT.Executer`:
    ```py
            # set environment variables
            self.env            = os.environ.copy()
            self.env.update({'DYLD_LIBRARY_PATH': Config.dyld_library_path})
            self.env.update({'LD_LIBRARY_PATH': Config.ld_library_path})
            self.env.update({'TZ': 'GMT-8'})
            self.env.update(extra_env)
    ```
- Reading variables in a format like `$exiv2` is not supported.

**Comment by clanmills** [comment]
1) You are supporting `Config.bin_dir`.  Excellent.
2) You fall back on PATH etc.  That's fine.
3) Substituting variables: "$exiv2 " => "../build/bin/exiv2" (or whatever's in suite.conf).  Can this be supported?

You have something really useful here.  We need to do some work to turn this from "What is this .....?." to "That is really well thought out and useful.".

**Comment by LeoHsiao1** [comment]
When I started porting bash_tests, I found that I only needed to read a few variables, such as `$exiv2`.
Therefore, I decided not to support reading variables in a format like `$exiv2`.
Instead, I made this the default behavior:
```py
>>> from system_tests import BT
>>> BT.Exec('exiv2 --help').args
['/mnt/d/1/exiv2/build/bin/exiv2', '--help']
>>> BT.Exec('exifprint --version').args
['/mnt/d/1/exiv2/build/bin/exifprint', '--version']
```
If the command you executed exists in the Config.bin_dir directory, it will be used and converted to an absolute path.

**Comment by clanmills** [comment]
**Much better**.  Well done.

```python
1264 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance/tests $ python3
Python 3.7.2 (v3.7.2:9a3ffc0492, Dec 24 2018, 02:44:43) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from system_tests import BT
>>> BT.Exec('exiv2 -vV')
exiv2 0.27.4.9
exiv2=0.27.4
...
version=Apple LLVM 12.0.0 (clang-1200.0.32.29)
date=Mar 10 2021
time=13:38:22
processpath=/Users/rmills/gnu/github/exiv2/0.27-maintenance/build/bin
executable=/Users/rmills/gnu/github/exiv2/0.27-maintenance/build/bin/exiv2
library=/Users/rmills/gnu/github/exiv2/0.27-maintenance/build/lib/libexiv2.0.27.4.9.dylib
```

```bash
1265 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance/tests $ env EXIV2_BINDIR=/usr/local/bin python3
Python 3.7.2 (v3.7.2:9a3ffc0492, Dec 24 2018, 02:44:43) 
>>> from system_tests import BT
>>> BT.Exec('exiv2 -vV')
exiv2 0.27.4.9
exiv2=0.27.4
..
executable=/usr/local/bin/exiv2
library=/usr/local/lib/libexiv2.0.27.4.9.dylib
...
```

**Comment by LeoHsiao1** [comment]
If the user has set EXIV2_BINDIR, that means that they want to use exiv2 in that directory. Therefore, I have increased the priority of EXIV2_BINDIR, ignoring PATH.

**Comment by clanmills** [comment]
That's enough for today.  I want to cut the grass and it's time for you to get to bed.

Tomorrow:

1. I will spend more time on README-TESTS.md
2. You'll investigate doing string substitution.
3. We will discuss "BT pipelines.".  I think they will be something like:

```python
>>> from system_tests import BT
>>> BT.grep('exec',BT.Exec('exiv2 -vV'))
```

Any easy and elegant syntax is fine.  Sip that juice and work hard.  We'll get there.

**Comment by clanmills** [comment]
Correct.  EXIV2_BINDIR is higher priority than PATH.

EXIV2_BINDIR was added by a student in 2010.  It was a good idea at the time.  It's not so smart when we give it more thought.  It should have been dealt with using `/usr/bin/env PATH=.....`.  However we're stuck with it.  It has to stay.

You have now given me **proof** that EXIV2_BINDIR executes the correct executable with the correct library.  Thank You and **Well Done**.

Dan's suite.conf magic is very elegant.  Let's get $variable string substitution into BT, and get pipelines working beautifully.  Wednesday, we'll discuss `diff` and other utilities.

Tomorrow, I will work on README-TESTS.md and test your HTTP_PORT fix on Cygwin and MinGW.

Tomorrow is another day.  Sleep Well.

**Comment by LeoHsiao1** [comment]
- I've just removed the default behavior of automatically converting absolute paths. Instead, add bin_dir to PATH.This is consistent with a normal shell script.
- `BT.config.init()` now has very little to do, so I've removed it.

**Comment by clanmills** [comment]
Here's my last words on Monday and then you're off to bed.

In the good old days of "bash_tests" written in bash, the person writing the test would mess about in the bash shell and develop the script.

In the great new days of 'BT' written in python, the person writing the test should mess about in the python shell and develop the script.  The aim is to provide a library of utilities (similar to bash) from python.

We've made very good progress today.  We will discuss and fix pipelines, help, utilities this week.

The tests you execute in 'tests/bash_tests/testcases.py are an application to test exiv2.  However the system_tests environment is a general purpose framework.

**Comment by clanmills** [comment]
@LeoHsiao1 There's something in https://github.com/Exiv2/exiv2/pull/1492/commits/b3b2caf5f070f7ae0623bc2740c4e3470280b68a that's breaking the build on Windows.  I think there's something horribly wrong with the PATH on Windows.

I don't know what you're doing.  So, I can't fix it.

I can see something that appears to be wrong.  You're determining the platform 'mingw' or 'cygwin' using:

```python
import os
platform=os.platform
```
That's not correct.  You should use system_tests.BT.verbose_version().get('platform').  Builds that have been done on Windows (msvc, Cygwin64 and MinGW/msys2) are using the python interpreter which believe sys.platform == 'win32'.  When exiv2 is built, it knows if he's being built for msvc, cygwin64, or mingw and exiv2 -vV outputs platform=windows | cygwin | mingw64 | mingw 32 | msys (see src/version.cpp#295).  A checkout that vv.get('platform').substr('mingw')==0 detect mingw. 

I'm sure there are other ways in python to detect cygwin and mingw.

Please revert the changes you made to README.md in https://github.com/Exiv2/exiv2/pull/1492/commits/05a600a84f755777801e445a590d88a419aa1251  "My aircraft".

**Comment by kmilos** [comment]
> I'm sure there are other ways in python to detect cygwin and mingw.

For msys and mingw this highly depends how you launched your session and what environment and python you're running as a result.

If you launched the "MSYS2 MSYS" terminal (POSIX emulation):

```
$ which python
/usr/bin/python

$ python -c "import os, sys; print(os.name); print(sys.platform); print(sys.version)"
posix
msys
3.8.7 (default, Jan 15 2021, 14:12:45)
[GCC 10.2.0]
```

(I think the only difference to cygwin is that you get `cygwin` instead of `msys`)

If OTOH you launched the "MSYS2 MinGW 64-bit" terminal (native Windows binaries):

```
$ which python
/mingw64/bin/python

$ python -c "import os, sys; print(os.name); print(sys.platform); print(sys.version)"
nt
win32
3.8.8 (default, Feb 20 2021, 07:16:03)  [GCC 10.2.0 64 bit (AMD64)]
```

Native Windows python output:

```
> python -c "import os, sys; print(os.name); print(sys.platform); print(sys.version)"
nt
win32
3.7.10 (default, Feb 26 2021, 13:06:18) [MSC v.1916 64 bit (AMD64)]
```

[How to tell between MinGW CPython and "official" Windows CPython](https://stackoverflow.com/a/51200002):

```python
if os.name == 'nt' and 'GCC' in sys.version
```

I think the moral of the story is: you shouldn't really be running msys python if you're trying to detect win32 API. msys == cygwin for all intents and purposes, i.e. those are POSIX environments.

**Comment by LeoHsiao1** [comment]
I'm worried that the verbose version that exiv2 outputs might be wrong, so I want to try other ways to determine the platform.
Thank you for your research. Can we adopt this way:
```py
import os
import sys

platform = sys.platform.lower() or 'unknown'    # It could be linux, win32, msys, cygwin, darwin, etc.
if os.name == 'nt' and 'GCC' in sys.version:
        platform = 'mingw'

```

**Comment by kmilos** [comment]
That's certainly better (i.e. complete), but the important thing to realize is that this only tells you what "platform" your python test suite is running on, not necessarily what your exiv2 binary is. My worry is that currently we seem to be testing the mingw binary on a msys platform...

**Comment by clanmills** [comment]
There are two quite different uses of PLATFORM.

When you are setting PATH, LD_LIBRARY_PATH or DYLD_LIBRARY_PATH, you must know the platform on which python is being used. sys.platform == 'win32' is perfect for that.  You're running on Windows.

When you set HTTP_PORT, use verbose_version().get('platform').

**Comment by LeoHsiao1** [comment]
Can we add a new test case?
```py
import os
import sys
import system_tests

platform = sys.platform.lower() or 'unknown'    # It could be linux, win32, msys, cygwin, darwin, etc.
if os.name == 'nt' and 'GCC' in sys.version:
        platform = 'mingw'

assert platform == system_tests.BT.verbose_version().get('platform')

```

**Comment by clanmills** [comment]
Gentlemen.  Let's not discuss HTTP_PORT at this time.  It's a tiny matter.  We need to get get the CI to be green.  The build started to fail with this submission.  https://github.com/Exiv2/exiv2/pull/1492/commits/b3b2caf5f070f7ae0623bc2740c4e3470280b68a

Can somebody revert that change, please.  I don't know or understand git.

**Comment by LeoHsiao1** [comment]
It would be two hours before I could go home and update the code.

**Comment by clanmills** [comment]
Stay Calm, Gentlemen.  The CI has been stable for months on msvc/cygwin/mingw.  The scripts to setup the cygwin and mingw is appveyor_mingw_cygwin.yml.  Luis set up the msvc/ci in appveyor.yml about 2 years ago.

Let's revert b3b2caf and @LeoHsiao1 can try again once the CI is green.

**Comment by clanmills** [comment]
Safe Journey Home, @LeoHsiao1 

I'm working on a redesign of clanmills.com.  So, I'm busy with something totally different.  Speak Later.

@kmilos Thank You for being involved.  Your assistance is appreciated.

**Review by clanmills — CHANGES_REQUESTED** [review]
I'm trying to withdraw my approval until this code is stable and better documented.

**Comment by kmilos** [comment]
Hope it's ok, I reverted to save @LeoHsiao1 some CI waiting time...

**Comment by LeoHsiao1** [comment]
I guess this doesn't work because the code that determines the operating system isn't in `BT.Config.init()`.

**Comment by clanmills** [comment]
Wonderful, @kmilos.  Thank You.   Please don't touch README.md - leave that 100% to me.

**Comment by kmilos** [comment]
> I guess this doesn't work because the code that determines the operating system isn't in `BT.Config.init()`.

Then it could actually be commit c1f0e2b

**Comment by clanmills** [comment]
@kmilos  and @LeoHsiao1  Good Work.

It feels a lot like Christmas - Red and Green Lights.

**Comment by LeoHsiao1** [comment]
I changed the code to identify the system through `platform.system()`.
It's not perfect, but this is how test cases work before you add `BT.verbose_version()`.
If CI is not restored, I need to analyze according to the error message.

**Comment by LeoHsiao1** [comment]
I see, the problem is because yesterday I removed the behavior that automatically converts the exiv2 command to an absolute path.
It appears that the PATH variable is not working properly, causing exiv2 to not be found.
```sh
======================================================================
ERROR: exiv2_test (bash_tests.testcases.TestCases)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\exiv2\tests\utils\exec.py", line 89, in run
    with subprocess.Popen(self.args, cwd=self.cwd, env=self.env,
  File "C:\msys64\mingw64\lib\python3.8\subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\msys64\mingw64\lib\python3.8\subprocess.py", line 1311, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "C:\projects\exiv2\tests\bash_tests\testcases.py", line 369, in exiv2_test
    out += re.sub(r'exiv2.*', 'exiv2 0.27.0.0 (__ bit build)', BT.Exec('exiv2 -u -V').stdout)
  File "C:\projects\exiv2\tests\utils\exec.py", line 72, in __init__
    self.run()
  File "C:\projects\exiv2\tests\utils\exec.py", line 98, in run
    raise RuntimeError('Failed to execute: {}'.format(self.args))
RuntimeError: Failed to execute: exiv2 -u -V
```

**Comment by clanmills** [comment]
@LeoHsiao1 Please revert the last couple of changes to return the CI to green.

Prepare your change, do a `git diff` and send it to me robin@clanmills.com with an explanation of what and why you have changed.  I will test it on Windows and macOS.  We also have a chat server.  There's a link in README.md

**Good News**.  I'm working on README-TESTS.md and that's going well.  This is going to be a very nice system.

**Comment by LeoHsiao1** [comment]
Please wait a few minutes while I solve the PATH problem so it doesn't come up again.

**Comment by clanmills** [comment]
@LeoHsiao1 Please let me fix the PATH for you.  Revert your code and send me a patch and explanation.

**Comment by LeoHsiao1** [comment]
I have re-converted the pending exiv2 command to an absolute path.
CI has been restored.

**Comment by LeoHsiao1** [comment]
Now we need to figure out why does PATH work on Linux and not on Windows.

**Comment by clanmills** [comment]
PATH works.  There's something wrong with how it's being set up.  Please explain what you are trying to do and I can fix it for you.  I'm totally lost in the complexity of your classes, __init__.py and other python magic.  You're making this more difficult than it really is.  We'll get this fixed.

**Comment by LeoHsiao1** [comment]
No, I debugged it several times, and it still didn't work when the child process was created:
```py
>>> import os
>>> import subprocess
>>> env = os.environ.copy()
>>> env.update({'PATH': os.path.normpath(r'D:\exiv2-0.27.2-2017msvc64\bin') + os.pathsep + env.get('PATH', '')})
>>> subprocess.Popen('exiv2', env=env)                  # PATH doesn't work
Traceback (most recent call last):
FileNotFoundError: [WinError 2]
>>> subprocess.Popen('exiv2', shell=True, env=env)      # PATH works
<subprocess.Popen object at 0x000002120A2AFF70>
exiv2: An action must be specified
exiv2: At least one file is required
Usage: exiv2 [ options ] [ action ] file ...

```

**Comment by clanmills** [comment]
I don't think the python version of bash_tests is ready to ship in v0.27.4.  This work should have been completed and working reliably by 2021-02-29.  The code in tests/bash_test is very complex, not working correctly and extremely volatile.

The documentation in tests/README-TESTS.md is a relic from tests/doc.md and very difficult to to understand.  I feel we need to rewrite tests/README-TESTS.md and document the new python classes required to implement `$ make bash_tests`.

I think we should release Exiv2 v0.27.4 with the existing bash scripts in test/*.sh.

We are planing to release Exiv2 v0.28 'main' in September 2021.  If @LeoHsiao1 wants to work on this in the months ahead, it can be brought into a robust state and will ship in the Fall.

**Comment by LeoHsiao1** [comment]
`system_tests.BT` were created to port shell test code to Python and have not been changed for half a year.
There were no plans to make it public, so no documentation was written.
If we were going to write new test cases using `system_tests.BT`, I would have to change some code to make it more generic.
